File: Source/SaveData/MCAFile.cpp
#include "MCAFile.h"
#include "ChunkNBT.h"
#include "Debug.h"
#include <fstream>
#include <iostream>
#include <string>
#include <cstdint>
#include <bitset>
#include <arpa/inet.h>

#include <cstdio>

// width/height in chunks of a region file:
static const constexpr int dimInChunks = 32;

// Loads data from a .mca file on construction.
MCAFile::MCAFile(std::filesystem::path filePath) : mcaPath(filePath)
{
    // read the region file's base coordinates from the file name:
    const std::string name(mcaPath.filename());
    const std::string numChars("-0123456789");
    const size_t xStart = name.find_first_of(numChars);
    const size_t xEnd = name.find('.', xStart);
    const size_t yStart = xEnd + 1;
    const size_t yEnd = name.find('.', yStart);
    if (xStart == std::string::npos || xEnd == std::string::npos
            || yEnd == std::string::npos)
    {
        std::cerr << "Can't parse coordinates from file " << mcaPath << ".\n";
        return;
    }
    int regionX = dimInChunks * std::stoi(name.substr(xStart, xEnd - xStart));
    int regionY = dimInChunks * std::stoi(name.substr(yStart, yEnd - yStart));

    std::ifstream regionFile(mcaPath, std::ios::binary);
    if (! regionFile.is_open())
    {
        std::cerr << "Failed to open " << mcaPath << "\n";
        return;
    }

    // Read a byte sequence:
    const auto readBytes = [this, &regionFile](char* buffer, size_t size)
    {
        if (regionFile.bad() || regionFile.eof())
        {
            std::cerr << mcaPath << ": file is in an invalid state.\n";
            return (size_t) 0;
            //exit(1);
        }
        regionFile.read(buffer, size);
        if (! regionFile || regionFile.gcount() != size)
        {
            std::cerr << mcaPath << ": only read " << regionFile.gcount()
                    << " bytes at " << regionFile.tellg() << ", expected "
                    << size << "\n";
            //exit(1);
        }
        return (size_t) regionFile.gcount();
    };

    // Read numBytes bytes of big-endian data into an unsigned integer:
    const auto readInt = [&regionFile, &readBytes](size_t numBytes)
    {
        uint64_t readValue = 0;
        if (numBytes > sizeof(readValue))
        {
            std::cerr << "Tried to read " << numBytes << " into a "
                    << sizeof(readValue) << " byte value!\n";
            return static_cast<uint64_t>(0);
            //exit(1);
        }
        std::vector<unsigned char> buffer;
        buffer.resize(numBytes);
        if(readBytes(reinterpret_cast<char*>(buffer.data()), numBytes)
                != numBytes)
        {
            return static_cast<uint64_t>(0);
        }
        size_t bitOffset = (numBytes - 1) * 8;
        for(const unsigned char byte : buffer)
        {
            readValue = readValue | (byte << bitOffset);
            bitOffset -= 8;
        }
        return readValue;
    };

    const constexpr int numChunks = dimInChunks * dimInChunks;
    const constexpr int bufferSize = numChunks * 4;
    unsigned char buffer [bufferSize];
    DBG("\nReading " << bufferSize << " bytes of index data from "
            << mcaPath << ":");
    static const constexpr int sectorSize = 4096; 
    const size_t fileSize = std::filesystem::file_size(mcaPath);
    DBG("File size: " << (fileSize / sectorSize) << " sectors, "
            << fileSize << " bytes.");
    size_t bytesRead = readBytes(reinterpret_cast<char*>(buffer), bufferSize);
    for (int i = 0; i < numChunks; i++)
    {
        const int chunkIndex = i * 4;
        if (chunkIndex >= bytesRead)
        {
            continue;
        }
        bool chunkLoaded = false;
        for (int cI = chunkIndex; cI < chunkIndex + 4; cI++)
        {
            if (buffer[cI] != 0)
            {
                chunkLoaded = true;
                break;
            }
        }
        if (chunkLoaded)
        {
            const int chunkX = regionX + (i % 32);
            const int chunkY = regionY + (i / 32);
            Point chunkPos = { chunkX, chunkY };

            // Find chunk data:
            DBG_V("Chunk " << (i + 1) << "/" << numChunks
                    << ", byte index " << chunkIndex << "/"
                    << ((numChunks - 1) * 4) << "\n");
            const int sectorOffset = (buffer[chunkIndex] << 16)
                    | (buffer[chunkIndex + 1] << 8)
                    | buffer[chunkIndex + 2];

            unsigned int sectorCount = buffer[chunkIndex + 3];
            unsigned int byteOffset = sectorOffset * sectorSize;
            if (byteOffset > fileSize)
            {
                std::cerr << "Chunk " << i << "/" << numChunks
                        << ", byte index " << chunkIndex
                        << ": Illegal offset past end of file: "
                        << byteOffset << " ("
                        << std::bitset<sizeof(byteOffset) * 8>(byteOffset)
                        << ", sector = " << sectorOffset << "/"
                        << (fileSize / sectorSize) << ")\n";

                continue;
            }
            DBG_V(i << ": Chunk " << chunkX << ", " << chunkY
                    << " data is " << sectorCount 
                    << " sector(s) at byte offset "
                    << (sectorOffset * sectorSize));
            regionFile.seekg(sectorOffset * sectorSize);
            if (! regionFile || regionFile.eof())
            {
                std::cerr << "Chunk " << (i + 1) << "/" << numChunks
                    << ", byte index " << chunkIndex << "/"
                    << ((numChunks - 1) * 4) << ": "
                    << "Failed to seek to offset "
                    << (sectorOffset * sectorSize) << " in file of size "
                    << fileSize << "\n";
                continue;
            }
            unsigned int chunkByteSize = readInt(4);
            unsigned char compressionType = readInt(1);
            size_t byteSectorCount = chunkByteSize / sectorSize;
            if ((chunkByteSize % sectorSize) > 0)
            {
                byteSectorCount++;
            }
            if (byteSectorCount > sectorCount)
            {
                std::cerr << i << ": Chunk " << chunkX << ", " << chunkY
                        << " at offset " << sectorOffset
                        << "/" << (fileSize / sectorSize) << ":\n"
                        << "Expected " << sectorCount << " sectors but found "
                        << byteSectorCount << "("
                        << std::bitset<32>(chunkByteSize) << ")\n";
                continue;
            }

            DBG_V(i << ": Chunk " << chunkX << ", " << chunkY
                    << " data is " << chunkByteSize << " bytes (" 
                    << (chunkByteSize / sectorSize) << ") sectors\n");

            std::vector<unsigned char> chunkData;
            chunkData.resize(chunkByteSize);
            size_t bytesRead = readBytes(
                    reinterpret_cast<char *>(chunkData.data()), chunkByteSize);
            if (bytesRead == chunkByteSize)
            {
                DBG_V(i << ": Chunk " << chunkX << ", " << chunkY
                        << ", " << bytesRead << "/" << chunkByteSize
                        << " bytes read.");
            }
            ChunkNBT nbtData(chunkData);
            loadedChunks.push_back(nbtData.getChunkData(chunkPos));
        }
    }
}


// Gets the coordinates of all loaded chunks stored in the file.
const std::vector<ChunkData>& MCAFile::getLoadedChunks()
{
    return loadedChunks;
}

File: src/main/java/com/centuryglass/chunk_atlas/savedata/MCAFile.java
/**
 * @file MCAFile.java
 * 
 * Reads data from a Minecraft region file.
 */

package com.centuryglass.chunk_atlas.savedata;

import com.centuryglass.chunk_atlas.config.LogConfig;
import com.centuryglass.chunk_atlas.util.ExtendedValidate;
import com.centuryglass.chunk_atlas.util.MapUnit;
import com.centuryglass.chunk_atlas.worldinfo.ChunkData;
import java.awt.Point;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.logging.Level;
import org.apache.commons.lang.Validate;

/**
 * Reads data from a Minecraft region file.
 */
public class MCAFile 
{
    private static final String CLASSNAME = MCAFile.class.getName();
    
    // width/height in chunks of a region file:
    private static final int DIM_IN_CHUNKS = 32;
    
    /**
     * Loads data from a .mca file on construction.
     *
     * @param mcaFile                 The Minecraft anvil region file to load.
     * 
     * @throws FileNotFoundException  If the file does not exist.
     */
    public MCAFile(File mcaFile) throws FileNotFoundException
    {
        final String FN_NAME = "MCAFile";
        ExtendedValidate.isFile(mcaFile, "Minecraft region file");
        loadedChunks = new ArrayList<>();
        // read the region file's base coordinates from the file name:
        Point regionPt = getChunkCoords(mcaFile);
        if (regionPt.x == -1 && regionPt.y == -1)
        {
            LogConfig.getLogger().logp(Level.WARNING, CLASSNAME, FN_NAME,
                    "Can't parse coordinates from file {0}.", mcaFile);
            return;
        }
        
        FileByteBuffer regionBuffer;
        try
        {
            regionBuffer = new FileByteBuffer(mcaFile);
        }
        catch (FileNotFoundException e)
        {
            LogConfig.getLogger().logp(Level.WARNING, CLASSNAME, FN_NAME,
                    "Failed to find file '{0}'.", mcaFile);
            return;
        }
        catch (IOException e)
        {
            LogConfig.getLogger().logp(Level.WARNING, CLASSNAME, FN_NAME,
                    "Error reading region file:", e);
            return;
        }
        
        LogConfig.getLogger().logp(Level.FINER, CLASSNAME, FN_NAME,
                    "Scanning region file {0}:", mcaFile.getAbsolutePath());
             
        // Read chunk data offsets within the file:
        int numChunks = DIM_IN_CHUNKS * DIM_IN_CHUNKS;
        final int sectorSize = 4096;
        List<Point> invalidChunks = new ArrayList<>();
        
        // If chunk loading fails, use this function to get chunk coordinates
        // from the chunk index:
        Function<Integer, Point> getPos = (index) ->
        {
            return new Point(regionPt.x + (index % 32),
                    regionPt.y + (index / 32));    
        };
        
        // Read all chunk offsets and extract chunk data:
        for (int i = 0; i < numChunks; i++)
        {
            long sectorOffset = 0;
            byte sectorCount = 0;
            try
            {
                sectorOffset = regionBuffer.readInt(3);
                sectorCount = regionBuffer.readByte();
            }
            catch (IndexOutOfBoundsException e)
            {
                LogConfig.getLogger().logp(Level.SEVERE, CLASSNAME, FN_NAME,
                        e.toString());
                // Invalid, out of bounds sector, skip it.
                loadedChunks.add(new ChunkData(getPos.apply(i),
                        ChunkData.ErrorFlag.BAD_OFFSET));
                invalidChunks.add(getPos.apply(i));
                regionBuffer.reset();
                continue;
            }
            if (sectorOffset == 0 && sectorCount == 0)
            {
                // That sector isn't loaded, skip it.
                loadedChunks.add(new ChunkData(getPos.apply(i),
                        ChunkData.ErrorFlag.CHUNK_MISSING));
                continue;
            }
            // Seek to offset and read chunk data:
            long byteOffset = sectorOffset * sectorSize;
            regionBuffer.mark();
            try
            {
                regionBuffer.setPos((int) byteOffset);
            }
            catch (IndexOutOfBoundsException e)
            {
                // Invalid, out of bounds sector, skip it.
                loadedChunks.add(new ChunkData(getPos.apply(i),
                        ChunkData.ErrorFlag.BAD_OFFSET));
                invalidChunks.add(getPos.apply(i));
                regionBuffer.reset();
                continue;
            }
            final int chunkByteSize = regionBuffer.readInt();
            if (chunkByteSize == 0)
            {
                continue;
            }
            regionBuffer.skipByte(); // compression type isn't needed
            byte[] chunkBytes = regionBuffer.readBytes(chunkByteSize);
            if (chunkBytes.length != chunkByteSize)
            {
                LogConfig.getLogger().logp(Level.WARNING, CLASSNAME, FN_NAME,
                        "Unexpected EOF: Read only {0} bytes, expected {1}.",
                        new Object[] { chunkBytes.length, chunkByteSize });
                invalidChunks.add(getPos.apply(i));
                continue;
            }
            ChunkNBT nbtData = new ChunkNBT(chunkBytes);
            ChunkData extractedData = nbtData.getChunkData();
            if (extractedData.getErrorType() != ChunkData.ErrorFlag.NONE)
            {
                extractedData = new ChunkData(getPos.apply(i),
                        extractedData.getErrorType());
            }
            loadedChunks.add(extractedData);
            regionBuffer.reset();
        }
        if (! invalidChunks.isEmpty())
        {
            LogConfig.getLogger().logp(Level.WARNING, CLASSNAME, FN_NAME,
                    "{0} chunks in region file '{1}' could not be loaded.",
                    new Object[] { invalidChunks.size(), mcaFile.getName() });
            LogConfig.getLogger().logp(Level.WARNING, CLASSNAME, FN_NAME,
                    "Invalid chunk coordinates: {0}",
                    invalidChunks.stream()
                    .map((Point point) -> {
                        Point blockCoords = new Point(regionPt.x + point.x,
                                regionPt.y + point.y);
                        blockCoords = MapUnit.convertPoint(blockCoords,
                                MapUnit.CHUNK, MapUnit.BLOCK);
                        return "(" + blockCoords.x + ", " + blockCoords.y + ")";
                     })
                    .reduce("", (String list, String point) -> {
                        return list + (list.isEmpty() ? "" : ", ") + point;
                    }));
        }
    }

    /**
     *  Finds a region file's upper left chunk coordinate from its file
     *         name.
     *
     * @param regionFile  A Minecraft region file.
     *
     * @return            The chunk coordinates, or null if the file name was
     *                    not properly constructed.
     */
    public static Point getChunkCoords(File regionFile)
    {
        Validate.notNull(regionFile, "Region file cannot be null.");
        final String name = regionFile.getName();
        final int xStart = 2;
        final int xEnd = name.indexOf(".", xStart);
        final int zStart = xEnd + 1;
        final int zEnd = name.indexOf(".", zStart);
        if (xEnd < 0 || zEnd < 0)
        {
            
            LogConfig.getLogger().logp(Level.WARNING, CLASSNAME, "getChunkCoords",
                    "chunk coords in region file '{0}' could not be loaded.",
                    new Object[] { regionFile });
            return null;
        }       
        return new Point(
                DIM_IN_CHUNKS * Integer.parseInt(name.substring(xStart, xEnd)),
                DIM_IN_CHUNKS * Integer.parseInt(name.substring(zStart,
                zEnd)));
    }

    /**
     *  Gets information about all loaded chunks stored in the file.
     *
     * @return  The list of chunk data objects.
     */
    public ArrayList<ChunkData> getLoadedChunks()
    {
        final String FN_NAME = "getLoadedChunks";
        if (loadedChunks.isEmpty())
        {
            LogConfig.getLogger().logp(Level.WARNING, CLASSNAME, FN_NAME,
                    "'{0}' had zero map chunks.", mcaFile);
        }
        return new ArrayList<>(loadedChunks);
    }
   
    private File mcaFile;
    private final ArrayList<ChunkData> loadedChunks;  
}
